google-cloud-spanner/lib/google/cloud/spanner/results.rb

```diff
@@ -198,6 +207,57 @@ def rows
 
         # rubocop:enable all
 
+        ##
+        # @private
+        # Checks if a request can be resumed by inspecting the resume token
+        def resumable? resume_token
+          !resume_token.nil? && !resume_token.empty?
+        end
+
+        ##
+        # @private
+        # Checks if a request can be retried. This is based on the error returned.
+        # Retryable errors are:
+        #   - Unavailable error
+        #   - Internal EOS error
+        #   - Internal RST_STREAM error
+        def retryable? err
+          err.instance_of?(GRPC::Unavailable) ||
+            (err.instance_of?(GRPC::Internal) && err.details.include?(EOS_INTERNAL_ERROR)) ||
+            (err.instance_of?(GRPC::Internal) && err.details.include?(RST_STREAM_INTERNAL_ERROR))
+        end
+
+        ##
+        # @private
+        # Resumes a request, by re-executing it with a resume token.
+        def resume_request! resume_token
```

Comment:
Why `exclamation mark` after the method? Do we modify any object in-place? Same for `retry_request! ` as well.

Reply:
Although it does not modify the object in-place it calls either `execute_streaming_sql` or `read` (in a nutshell performs side effects - not pure function). This is the reasoning behind this choice. Let me know if you'd like me to remove it.

Reply:
But that means even `client.execute_query` should also have `!` at the end. To me,  I would think adding `!` means that it modifies an object in-place.

Reply:
It's a private method. I am okay with any approach, but it would be better to have a consistent convention in client lib.

Reply:
No worries, I will remove the bangs. I agree with you that having a consistent convention is more important.

Reply:
Done.